packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,478 @@
+	ref: LocalReferencePosition,
+	segmentIndex: number,
+	startIntervals: { segmentIndex: number; interval: SequenceInterval }[],
+	endIntervals: { segmentIndex: number; interval: SequenceInterval }[],
+) {
+	if (refTypeIncludesFlag(ref.refType, ReferenceType.RangeBegin)) {
+		const interval = ref.properties?.interval;
+		if (interval && interval instanceof SequenceInterval) {
+			startIntervals.push({ segmentIndex, interval });
+			return true;
+		}
+	} else if (refTypeIncludesFlag(ref.refType, ReferenceType.RangeEnd)) {
+		const interval = ref.properties?.interval;
+		if (interval && interval instanceof SequenceInterval) {
+			endIntervals.push({ segmentIndex, interval });
+			return true;
+		}
+	}
+	return false;
+}
+
+function addLocalRefPlaceholders(string: SharedString, ref: LocalReferencePosition) {
+	return {
+		stay: string.createLocalReferencePosition(
+			ref.getSegment() as SharedStringSegment,
+			ref.getOffset(),
+			ReferenceType.StayOnRemove,
+			undefined,
+		),
+		slide: string.createLocalReferencePosition(
+			ref.getSegment() as SharedStringSegment,
+			ref.getOffset(),
+			ReferenceType.SlideOnRemove,
+			undefined,
+		),
+	};
+}
+
+export function appendToMergeTreeDeltaRevertibles(
+	string: SharedString,
+	deltaArgs: IMergeTreeDeltaCallbackArgs,
+	revertibles: SharedStringRevertible[],
+) {
+	if (deltaArgs.deltaSegments.length === 0) {
+		return;
+	}
+	if (deltaArgs.operation === MergeTreeDeltaType.REMOVE) {
+		const startIntervals: { segmentIndex: number; interval: SequenceInterval }[] = [];
+		const endIntervals: { segmentIndex: number; interval: SequenceInterval }[] = [];
+		let position: number | undefined;
+
+		// find interval endpoints in each segment
+		for (let i = 0; i < deltaArgs.deltaSegments.length; i++) {
+			const refs = deltaArgs.deltaSegments[i].segment.localRefs;
+			if (refs !== undefined) {
+				for (const ref of refs) {
+					addIfIntervalEndpoint(ref, i, startIntervals, endIntervals);
+				}
+
+				// get current position of deletion to search intervals later
+				const segPosition = string.getPosition(deltaArgs.deltaSegments[i].segment);
+				if (position === undefined) {
+					position = segPosition;
+				} else {
+					assert(
+						position === segPosition,
+						"Discontinuous local removals are not supported",
+					);
+				}
+			}
+		}
+
+		if (
+			position !== undefined &&
+			position !== -1 &&
+			(startIntervals.length > 0 || endIntervals.length > 0)
+		) {
+			// make revertible
+			if (startIntervals.length > 0 || endIntervals.length > 0) {
+				const removeRevertibles: MergeTreeDeltaRevertible[] = [];
+				appendToMergeTreeRevertibles(string, deltaArgs, removeRevertibles);
+				assert(
+					removeRevertibles.length === 1,
+					"Remove revertible should be a single delta",
+				);
+
+				const revertible: TypedRevertible<typeof IntervalEventType.SEQUENCEREMOVE> = {
+					event: IntervalEventType.SEQUENCEREMOVE,
+					intervals: [],
+					mergeTreeRevertible: removeRevertibles[0],
```

Comment:
ok. i see how the index tracking work for now. you depend on the current behavior where each removed segment maps to a single local reference at remove time, so we insert the exact same number of references. I was actually thinking about changing that to improve the merge behavior, and letting revertible local references split, which would break that mapping. Probably not something we need to solve immediatly.

something that could work is to instead store the offest within the removed range, then on re-insert you treat the segements as a contigous for per offset applications. Specficially you just sum up the cached lengths of all the segments, and then for each range offset, you find which seg, and seg off it applies to, then ask the share string to get the the real position in the tree.

Reply:
ok. i see how the index tracking work for now. you depend on the current behavior where each removed segment maps to a single local reference at remove time, so we insert the exact same number of references. I was actually thinking about changing that to improve the merge behavior, and letting revertible local references split, which would break that mapping. Probably not something we need to solve immediatly.

something that could work is to instead store the offest within the removed range, then on re-insert you treat the segements as a contigous for per offset applications. Specficially you just sum up the cached lengths of all the segments, and then for each range offset, you find which seg, and seg off it applies to, then ask the share string to get the the real position in the tree.